Melt enhance

tdpetrou opened the pull request:
- [ x] closes #17676
- [ x] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

tdpetrou committed:
BUG: coerce pd.wide_to_long suffixes to numeric

tdpetrou committed:
ENH: simultaneous melting

jreback commented on pandas/core/reshape/reshape.py:
```diff
@@ -715,71 +716,241 @@ def _convert_level_number(level_num, columns):
-def melt(frame, id_vars=None, value_vars=None, var_name=None,
-         value_name='value', col_level=None):
+def _melt(frame, id_vars=None, value_vars=None, var_name=None,
+          value_name='value', col_level=None, stubnames=False,
```
i would be fine with another module called melt.py
which contained all of the code for melt; 

that way helper functions can be broken out and impl is understandable

tdpetrou replied:
I created melt.py and test_melt.py. I moved melt out of reshape.py and corrected some imports. I heavily commented the code as well. I do need at least 20 more tests as there is a huge range of possibilities now with this new implementation.

jreback commented on pandas/tests/reshape/test_reshape.py:
```diff
@@ -33,6 +34,33 @@ def setup_method(self, method):
         self.df1.columns.names = ['CAP', 'low']
 
+        self.df2 = DataFrame(
+            {'City': ['Houston', 'Austin', 'Hoover'],
```
also move to test_melt.py (all melt related)

tdpetrou committed:
added lots of comments

jreback commented:
@tdpetrou I wouldn't go too far atm. This needs quite a lot of review. I am not real happy with all of these arguments.

tdpetrou commented:
Yea, the wide_to_long functionality can be nixed. It doesn't even exactly replicate it and then we'd be back to the same function signature

tdpetrou commented:
If you really wanted to incorporate wide_to_long functionality and minimize arguments, you could have a single extra parameter prefix that takes either a string or a two-item tuple with the first item being the separator.

tdpetrou commented:
@jreback Can we review this? This new melt function adds a ton of new functionality that I think would be very beneficial.

jreback commented:
if you can rebase would be great

jorisvandenbossche commented:
@tdpetrou I would like to review this PR, but it is very difficult to see what has changed or what are the added enhancements to melt. Therefore, can you update the docstring, add a whatsnew note and add narrative docs in reshaping.rst ?
I agree with moving the implementation to a separate melt.py module, but for reviewing purposes I maybe would like to keep it where it was, and only afterwards move it (so not refactor + move at the same time).

> Yea, the wide_to_long functionality can be nixed. It doesn't even exactly replicate it and then we'd be back to the same function signature

For reviewing and discussing purposes, it might also be better to keep it in two separate PRs (the ability to melt on multiple columns (the referenced issue) and the introduction of wide_to_long similar functionality). But I don't know to what extent the implementation of the one is needed for the other? (how independent both additions are?)

jreback commented:
This proposed API is very complicated. i would be much more in favor of something much simpler. Having a single function do a single thing is much much more intutive. A possible resolution is to introduce a lazy object with methods itself, akin to groupby. e.g.

```
r = df.reshape(.....)
r.melt()
r.to_long()
```

or somesuch, with each function performing 1 action

tdpetrou commented:
To first address the two-step function. I agree with @jorisvandenbossche and don't think that is necessary. Although, the `wide_to_long` functionality can be captured with a single additional parameter, let's keep things simple and just drop it completely from this PR. 

So, I will:

- Keep melt in its original file reshape.py
- Remove the wide to long functionality - meaning no extra parameters
- Change the whatsnew
- Add notes to reshaping.rst

TomAugspurger commented:
@tdpetrou can you take a look at https://github.com/pandas-dev/pandas/pull/17459 and see how those would conflict / overlap with eachother?

I haven't had a chance to go through the changes here in detail yet.

tdpetrou commented:
I think everything is updated and rebased appropriately. I had to delete a few changes #17628 that were mixed in here as well. I made a little notebook on the enhancement as well (https://nbviewer.jupyter.org/github/tdpetrou/Machine-Learning-Books-With-Python/blob/master/melt%20enhancemenet.ipynb)

@TomAugspurger I don't think it will be much of a problem to keep the index. It'll just be like another entry to`id_vars`

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -13,7 +13,7 @@ version.
 ~~~~~~~~~~~~
 
--
+- Simultaneous melting of independent groups of columns is now possible with ``melt``.
```
you will need a full example, pls pointers to the docs. but let's settle on an API.

tdpetrou replied:
Added Sentence to 'highlights' section and two examples to 'New Features' section.

API will be backwards compatiable. 
* `value_vars` can take a list of lists
* `value_name` and `var_name` can take a list of new column names

@TomAugspurger mentioned the addition of `keep_index` parameter. This can easily be added.

`wide_to_long` funcitonality can be exactly duplicated (and more) with a single additional parameter that takes a tuple of `suffix` and `sep` but this might be too much for one function. 

Regardless, `wide_to_long` can be refactored to be faster.

This new melt enhancement makes lreshape obsolete (other than it being faster)

jreback commented on pandas/core/frame.py:
```diff
@@ -4628,6 +4629,31 @@ def unstack(self, level=-1, fill_value=None):
+
+    Simultaneously melt multiple groups of columns:
+
+    >>> df2 = pd.DataFrame({'City': ['Houston', 'Miami'],
```
simple examples first

tdpetrou replied:
This is as simple as it gets. One id column, two column groups of each length two and only two rows of data.

jreback commented:
so, this is pretty much impossible to review as is. I would first do a simple PR which moves the implementation of .melt to a separate module. This would be a simple copy/paste. Then you can make this implementation much more simple, but for example not in-lining functions inside melt itself, rather defining them in the module.

tdpetrou committed:
updated melt docs and put melt in own module

tdpetrou commented:
@jreback 
I ....
* put melt in its own module
* removed any nested functions
* made a separate tests module
* added a couple examples to whatsnew
* simplified an example and added one in the reshaping docs

jreback commented on doc/source/reshaping.rst:
```diff
@@ -265,24 +265,59 @@ the right thing:
+   df = pd.DataFrame({'State': ['Texas', 'Florida', 'Alabama'],
+                      'Mango':[4, 10, 90],
+                      'Orange': [10, 8, 14], 
+                      'Watermelon':[40, 99, 43]},
```
generaly like to keep docs to 80 chars or less (readibility)

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -8,12 +8,50 @@ deprecations, new features, enhancements, and performance improvements along
+
+Simultaneous unpivoting of independent groups of columns with ``melt``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Previously, ``melt`` was only able to unpivot a single group of columns. This was done by passing all the column names in the group as a list to the ``value_vars`` parameter.
```
add the issue number(s)

jreback commented on pandas/core/frame.py:
```diff
@@ -4554,13 +4554,14 @@ def unstack(self, level=-1, fill_value=None):
-    var_name : scalar
-        Name to use for the 'variable' column. If None it uses
+        Column(s) to unpivot. If list of lists, simultaneously unpivot
+        each sublist into its own variable column. If not specified, uses all
```
add a mention where this changes as appropriate (meaning in 0.22.0) in the doc-string

jreback reviewed (changes requested):
I want a separate PR entirely which simply moves melt to melt.py. don't add anything (except changes that are required in .api for example) or testing to make sure it all passes green. we will then merge that. you can rebase this pr on top. this way its easy to see what you actually changed.

tdpetrou commented:
@jreback Getting back to this now. I put melt in its own module as requested #18148

pep8speaks commented:
Hello @tdpetrou! Thanks for updating the PR.





 - In the file [`pandas/tests/reshape/test_melt.py`](https://github.com/pandas-dev/pandas/blob/755c3db960fa34db66efbc0c3787b64acfc60a92/pandas/tests/reshape/test_melt.py), following are the PEP8 issues :

> [Line 705:48](https://github.com/pandas-dev/pandas/blob/755c3db960fa34db66efbc0c3787b64acfc60a92/pandas/tests/reshape/test_melt.py#L705): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

tdpetrou commented:
@jreback I just rebased this and would like to work on it

tdpetrou committed:
added newline

tdpetrou committed:
0.21 whatsnew back to original

jreback commented:
ok. this is doing a lot of things. I would appreciate to break this apart into separate distinct changes. The code needs some work. I can comment, but easier to do it in smaller pieces.

tdpetrou commented:
@jreback There is basically one piece, now that the wide_to_long functionality has been removed. Essentially, you pass a list of lists into `value_vars` to simultaneously melt those sets of columns. 

I suppose you could break it up into 3 pieces
* Make only original `melt` work with my new way
* Add in MultiIndex support - this handles all varieties of string/integer MultiIndex
* Finally add in simultaneously melt

I think I can break it up like this without too much effort. What do you think?

jreback commented:
closing as stale. ping if you want to update.
